File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n=nums.size();
        int first=0;
        int second=0;
        int prefix=0;
        int ans=0;
        unordered_map<int,int>map;
        map[0]=1;
        while(second<n){
            prefix+=nums[second];
            int target=prefix-k;
            if (map.count(target)){
                ans+=map[target];
                first++;
            }
            map[prefix]++;
            second++;
        }
        return ans;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int prefix=0;
        
        int total=0;
  
        HashMap<Integer,Integer>mp=new HashMap<>();
           mp.put(0,1);

        for(int i=0;i<nums.length;i++){
            prefix +=nums[i];

           int target=prefix-k;

            if(mp.containsKey(target)){
                total +=mp.get(target);
            } 
            mp.put(prefix, mp.getOrDefault(prefix, 0) + 1);
        }
        return total;
    }
}